lib/theory/src/Theory/Constraint/System/JSON.hs

```diff
@@ -0,0 +1,394 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- |
+-- Copyright   : (c) 2016 Dominik Schoop
+-- License     : GPL v3 (see LICENSE)
+--
+-- Maintainer  : Dominik Schoop dominik.schoop@hs-esslingen.de
+-- Portability : GHC only
+--
+-- Conversion of the graph part of a sequent (constraint system) to a JSON graph format
+
+{-
+   ADVISE:
+   - DO NOT USE sequentToJSONPretty IN OPERATIONAL MODE.
+     Remember to replace sequentToJSONPretty with sequentToJSON in getTheoryGraphR in Handler.hs.
+   - The time consuming pretty printing of facts and terms can be disabeled by a parameter of
+     sequentToJSONPretty and sequentToJSON. Currently, everything is pretty with sequentToJSONPretty
+     and nothing is pretty with sequentToJSON.
+
+   TO DO:
+   - use console parameter to control whether DOT or JSON is used 
+   - generate JSON in non-interactive mode
+   - make it work for observational equivalence
+   - encode historic information in the graph: sequence of nodes added
+
+   OPEN PROBLEMS:
+   - JSON encodePretty converts < and > to "\u003c" and "\u003e"
+     ghastly postprocessing has been applied to sequentToJSONPretty
+-}
+
+module Theory.Constraint.System.JSON (
+    sequentToJSON,                     
+    writeSequentAsJSONToFile,
+    sequentToJSONPretty,
+    writeSequentAsJSONPrettyToFile
+  ) where
+import           Extension.Data.Label       as L (get)
+import           Data.Aeson
+import           Data.Aeson.TH
+import           Data.Aeson.Encode.Pretty   -- to do pretty printing of JSON
+import qualified Data.Map                   as M
+import           Data.Maybe
+import qualified Data.Set                   as S
+import qualified Data.ByteString.Lazy.Char8 as BC (unpack)
+
+import           Text.PrettyPrint.Class     -- for Doc and the pretty printing functions 
+import           Theory.Constraint.System   
+import           Theory.Model
+
+------------------------------------------------------------------------------
+-- Data structure for JSON graphs                                           --
+-- adapted from https://github.com/jsongraph/json-graph-specification       --
+------------------------------------------------------------------------------
+
+-- term in a JSON graph node, either in a fact, an action or an unsolvedActionAtom
+data JSONGraphNodeTerm
+  = Const { jgnConst :: String }
```

Comment:
The easiest is to just drop the field  names from the constructors of
sum-types. If the types of the constructor arguments are different, then
positional indexing is just as fail-safe as named access.

If not, then you'd ideally introduce actual record types for each kind of
constructor. That's a bit cumbersomr though. Our style-guide should have a
section about a scalable naming-scheme for this case.

dschoop <notifications@github.com> schrieb am Fr., 12. Feb. 2016 12:08 PM:

> In lib/theory/src/Theory/Constraint/System/JSON.hs
> <https://github.com/tamarin-prover/tamarin-prover/pull/168#discussion_r52729184>
> :
>
> > +import           Data.Maybe
> > +import qualified Data.Set                   as S
> > +import qualified Data.ByteString.Lazy.Char8 as BC (unpack)
> > +
> > +import           Text.PrettyPrint.Class     -- for Doc and the pretty printing functions
> > +import           Theory.Constraint.System
> > +import           Theory.Model
> > +
> > +------------------------------------------------------------------------------
> > +-- Data structure for JSON graphs                                           --
> > +-- adapted from https://github.com/jsongraph/json-graph-specification       --
> > +------------------------------------------------------------------------------
> > +
> > +-- term in a JSON graph node, either in a fact, an action or an unsolvedActionAtom
> > +data JSONGraphNodeTerm
> > +  = Const { jgnConst :: String }
>
> I think I understand the problem but do not know how to solve it elegantly
> with Aeson. Can you give me a hint?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tamarin-prover/tamarin-prover/pull/168/files#r52729184>
> .
>